packages/ra-core/src/controller/field/useReferenceManyFieldController.ts

```diff
@@ -84,13 +70,140 @@ const useReferenceManyFieldController = ({
-}: Options): ReferenceManyProps => {
-    const referenceId = get(record, source);
+    page: initialPage,
+    perPage: initialPerPage,
+    sort: initialSort = { field: 'id', order: 'DESC' },
+}: Options): ListControllerProps => {
     const notify = useNotify();
-    const { data, ids, total, loading, loaded } = useGetManyReference(
+
+    const stateParams = useRef({
+        page: initialPage,
+        perPage: initialPerPage,
+        sort: initialSort,
+        filter,
+    });
+
+    // pagination logic
+    const [page, setPage] = useSafeSetState<number>(initialPage);
+    const [perPage, setPerPage] = useSafeSetState<number>(initialPerPage);
+    // handle pagination props change
+    useEffect(() => {
+        if (initialPage !== stateParams.current.page) {
+            stateParams.current.page = initialPage;
+            setPage(initialPage);
+        }
+    }, [initialPage, setPage]);
+    useEffect(() => {
+        if (initialPerPage !== stateParams.current.perPage) {
+            stateParams.current.perPage = initialPerPage;
+            setPerPage(initialPerPage);
+        }
+    }, [initialPerPage, setPerPage]);
+
+    // sort logic
+    const [sort, setSortObject] = useSafeSetState<Sort>(initialSort);
+    const setSort = useCallback(
+        (field: string, order: string = 'ASC') => {
+            setSortObject(previousState => ({
+                field,
+                order:
+                    field === previousState.field
+                        ? previousState.order === 'ASC'
+                            ? 'DESC'
+                            : 'ASC'
+                        : order,
+            }));
+            setPage(1);
+        },
+        [setPage, setSortObject]
+    );
+    // handle sort props change
+    useEffect(() => {
+        if (!isEqual(initialSort, stateParams.current.sort)) {
+            stateParams.current.sort = initialSort;
+            setSortObject(initialSort);
+        }
+    }, [initialSort, setSortObject]);
+
+    // selection logic
+    const [selectedIds, setSelectedIds] = useSafeSetState<Identifier[]>([]);
+    const onSelect = useCallback(
+        (newIds: Identifier[]) => {
+            setSelectedIds(newIds);
+        },
+        [setSelectedIds]
+    );
+    const onToggleItem = useCallback(
+        (id: Identifier) => {
+            setSelectedIds(previousState => {
+                const index = previousState.indexOf(id);
+                if (index > -1) {
+                    return [
+                        ...previousState.slice(0, index),
+                        ...previousState.slice(index + 1),
+                    ];
+                } else {
+                    return [...previousState, id];
+                }
+            });
+        },
+        [setSelectedIds]
+    );
+    const onUnselectItems = useCallback(() => {
+        setSelectedIds([]);
+    }, [setSelectedIds]);
+
+    // filter logic
+    const [displayedFilters, setDisplayedFilters] = useSafeSetState<{
```

Comment:
useFilterState ?

Reply:
Unfortunately, this hook doesn't handle the complex logic of setting filters, displayed filters, and page at the same time, so I con't use it in this case.